### PR TITLE
feat(Firma con IO): [SFEQS-1060] Add qtsp clauses and document preview screens

### PR DESCRIPTION
**Pull request by hevelius**
## Short description
In this PR, a screen has been added to show the QTSP clauses that the user must accept. In a clause there should be one or more links to a pdf document. Some clause other than the regular URL may contain a TAG_LINK.

```
{
  clauses: [
    {
      text: This is the clause text [text to link](https://normalLink/example.pdf) with more links [text to link](@TAG_LINK)
    }
  ]
}
```

You can map TAG_LINKS to replace them with other URLs. TAG_LINKS was introduced because in the FCI signature flow some clauses have a link to be replaced dynamically with other contained user data. (This could be a temporarily solutions because changes could be by QTSP)

https://user-images.githubusercontent.com/49342144/205942495-684d9397-4edc-482d-a7c3-6857d4ab89fc.mov

## List of changes proposed in this pull request
- Added `FciQtspClausesScreen`
- Added `FciDocumentPreviewScreen`
- Added `DocumentViewer` component
- Added `QtspSignatureField` and `LinkedText` components
- Added a saga to handle the download of pdf document

## How to test
* if not yet merged get this branch https://github.com/pagopa/io-dev-api-server/tree/SFEQS-1082-fci-endpoints from io-dev-server-api
* enable FCI_ENABLED ff
* follow signature flow
* on the QTSP clauses screen try to tap on a link to show related pdf in preview
* the continue button it should be activate when all checkbox are selected

### CODE REVIEW

File: ts/features/fci/components/DocumentViewer.tsx [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+      }, I18n.t("features.mvl.details.attachments.pdfPreview.singleBtn"))}
+    />
+  ) : (
+    <FooterWithButtons
+      type={"ThreeButtonsInLine"}
+      leftButton={{
+        bordered: true,
+        primary: false,
+        onPress: () => {
+          onShare?.();
```
Comment (francescopersico) [inline ts/features/fci/components/DocumentViewer.tsx:67]: I can't get the point of displaying a button that do nothing on press. Do we probably need to hide it?

File: ts/features/fci/components/DocumentViewer.tsx [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+            );
+          });
+        },
+        title: I18n.t("global.buttons.share")
+      }}
+      midButton={{
+        bordered: true,
+        primary: false,
+        onPress: () => {
+          onDownload?.();
```
Comment (francescopersico) [inline ts/features/fci/components/DocumentViewer.tsx:82]: As above

File: ts/features/fci/components/DocumentViewer.tsx [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                I18n.t(
+                  "features.mvl.details.attachments.pdfPreview.errors.saving"
+                )
+              );
+            });
+        },
+        title: I18n.t("features.mvl.details.attachments.pdfPreview.save")
+      }}
+      rightButton={confirmButtonProps(() => {
+        onOpen?.();
```
Comment (francescopersico) [inline ts/features/fci/components/DocumentViewer.tsx:112]: As above

File: ts/features/fci/components/DocumentViewer.tsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  const dispatch = useIODispatch();
+  const navigation = useNavigation();
+  const fciDownloadSelector = useIOSelector(fciDownloadPreviewSelector);
+  const fciDownloadPath = useIOSelector(fciDownloadPathSelector);
+
+  React.useEffect(() => {
+    dispatch(fciDownloadPreview.request({ url: documentUrl }));
+  }, [documentUrl, dispatch]);
+
+  const LoadingComponent = () => (
```
Comment (francescopersico) [inline ts/features/fci/components/DocumentViewer.tsx:147]: Can we move this outside the function?

File: ts/features/fci/components/DocumentViewer.tsx [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+      goBack={true}
+      customGoBack={customGoBack}
+      contextualHelp={emptyContextualHelp}
+      headerTitle={I18n.t("features.mvl.details.attachments.pdfPreview.title")}
+    >
+      <SafeAreaView
+        style={styles.container}
+        testID={"FciDocumentPreviewScreenTestID"}
+      >
+        {pipe(fciDownloadPath, S.isEmpty) === false && (
```
Comment (francescopersico) [inline ts/features/fci/components/DocumentViewer.tsx:185]: Is this pipe useful? Isn't S.isEmpty(fciDownloadPath) enough?
Is correct that nothing is rendered when the fciDownloadPath is empty?

File: ts/features/fci/components/LinkedText.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    onPress: (holder: string) => void
+  ) => {
+    const parts = getMatchedLinks(text);
+
+    // a function to render the link
+    const getTextWithLinkComponent = (matched: string, index: number) => {
+      const splitted = matched.split(matchMarkdownLink);
+      const textToBeLinked = splitted[1];
+      const url = splitted[2];
+      // TODO: add support for other TAG links
```
Comment (francescopersico) [inline ts/features/fci/components/LinkedText.tsx:51]: Can we reference a issue for this todo?

File: ts/features/fci/components/QtspClauseListItem.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import { IOStyles } from "../../../components/core/variables/IOStyles";
+import { IOColors } from "../../../components/core/variables/IOColors";
+import TouchableDefaultOpacity from "../../../components/TouchableDefaultOpacity";
+import { QtspClause } from "../../../../definitions/fci/QtspClause";
+import { fciQtspFilledDocumentUrlSelector } from "../store/reducers/fciQtspFilledDocument";
+import LinkedText from "./LinkedText";
+
+type Props = {
+  clause: QtspClause;
+  value?: boolean;
```
Comment (francescopersico) [inline ts/features/fci/components/QtspClauseListItem.tsx:14]: Isn't value a bit generic? What about "checked"?

File: ts/features/fci/components/LinkedText.tsx [hunk 8]
```diff
@@ -0,0 +1,92 @@
+import * as React from "react";
+import * as O from "fp-ts/lib/Option";
+import { H4 } from "../../../components/core/typography/H4";
+import { Link } from "../../../components/core/typography/Link";
+import { WithTestID } from "../../../types/WithTestID";
+
+type Props = WithTestID<{
+  text: string;
+  url: string;
```
Comment (francescopersico) [inline ts/features/fci/components/LinkedText.tsx:9]: It's not clear for an external user of the component what this "url" means.
What about an object with **tag** as key and **replacement** as value?

File: ts/features/fci/screens/valid/FciQtspClausesScreen.tsx [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+import { H4 } from "../../../../components/core/typography/H4";
+import { Link } from "../../../../components/core/typography/Link";
+
+const styles = StyleSheet.create({
+  paddingText: {
+    paddingLeft: 4
+  }
+});
+const FciQtspClausesScreen = () => {
+  const [isClausesChecked, setIsClausesChecked] = React.useState(0);
```
Comment (francescopersico) [inline ts/features/fci/screens/valid/FciQtspClausesScreen.tsx:31]: "is" prefix is used for boolean values. try to find a name that express a quantity/number.

File: ts/features/fci/store/actions/index.ts [hunk 10]
```diff
@@ -47,6 +47,20 @@ export const fciSigningRequest = createAsyncAction(
...
+ * asycn action to download file
+ */
+export const fciDownloadPreview = createAsyncAction(
+  "FCI_DOWNLOAD_PREVIEW_REQUEST",
+  "FCI_DOWNLOAD_PREVIEW_SUCCESS",
+  "FCI_DOWNLOAD_PREVIEW_FAILURE",
+  "FCI_DOWNLOAD_PREVIEW_CANCEL"
+)<{ url: string }, { path: string }, NetworkError, void>();
+
+export const fciDownloadPreviewCancel = createStandardAction(
```
Comment (francescopersico) [inline ts/features/fci/store/actions/index.ts:60]: Isn't the cancel already defined on the action above?

**Review by francescopersico — CHANGES_REQUESTED** [review]

File: ts/features/fci/components/DocumentViewer.tsx (hunk 1 above)
Comment (hevelius) [reply ts/features/fci/components/DocumentViewer.tsx:67]: Ok. I'll try to remove this code. This code is a copy of the component used in PN where only the mode to get the url is changed.

File: ts/features/fci/components/DocumentViewer.tsx (hunk 5 above)
Comment (hevelius) [reply ts/features/fci/components/DocumentViewer.tsx:185]: At the beginning of the render there is a check on `fciDownloadSelector` where `fciDownloadPath` come from. Maybe this check could be unuseful. I'll try to remove it.

File: ts/features/fci/components/LinkedText.tsx (hunk 8 above)
Comment (hevelius) [reply ts/features/fci/components/LinkedText.tsx:9]: I'll change the "url" name to a clearer one for now. There will be a refactor dependent on this coming soon related to the https://pagopa.atlassian.net/browse/SFEQS-1230

File: ts/features/fci/store/actions/index.ts (hunk 10 above)
Comment (hevelius) [reply ts/features/fci/store/actions/index.ts:60]: Yep. It's my fault already fixed in the #4239 I update this one. Thx!

**Review by francescopersico — APPROVED** [review]

**Pull request merged**